pkg/credentialprovider/plugin/plugin.go

```diff
@@ -727,89 +790,39 @@ func mergeEnvVars(sysEnvVars, credProviderVars []string) []string {
 	return mergedEnvVars
 }
 
-// generateServiceAccountCacheKey generates the serviceaccount cache key to be used for
-// 1. constructing the cache key for the service account token based plugin in addition to the actual cache key (image, registry, global).
-// 2. the unique key to use singleflight for the plugin in addition to the image.
-func generateServiceAccountCacheKey(serviceAccountNamespace, serviceAccountName string, serviceAccountUID types.UID, saAnnotations map[string]string) (string, error) {
-	b := cryptobyte.NewBuilder(nil)
-	b.AddUint16LengthPrefixed(func(b *cryptobyte.Builder) {
-		b.AddBytes([]byte(serviceAccountNamespace))
-	})
-	b.AddUint16LengthPrefixed(func(b *cryptobyte.Builder) {
-		b.AddBytes([]byte(serviceAccountName))
-	})
-	b.AddUint16LengthPrefixed(func(b *cryptobyte.Builder) {
-		b.AddBytes([]byte(serviceAccountUID))
-	})
-
-	// add the length of annotations to the cache key
-	b.AddUint32(uint32(len(saAnnotations)))
+func generateServiceAccountCacheKey(params cacheKeyParams) (string, error) {
+	builder := newCacheKeyBuilder().
+		addString(params.namespace).
+		addString(params.serviceAccountName).
+		addString(string(params.serviceAccountUID)).
+		addAnnotations(params.saAnnotations)
 
-	// Sort the annotations by key to ensure the cache key is deterministic
-	keys := sets.StringKeySet(saAnnotations).List()
-	for _, k := range keys {
-		b.AddUint16LengthPrefixed(func(b *cryptobyte.Builder) {
-			b.AddBytes([]byte(k))
-		})
-		b.AddUint16LengthPrefixed(func(b *cryptobyte.Builder) {
-			b.AddBytes([]byte(saAnnotations[k]))
-		})
+	if needsPodInfo(params.cacheType) {
```

Comment:
let's just add these if they made their cache type `Pod`

`if cacheType == kubeletconfig.PodServiceAccountTokenCacheType`

 if they made it `Token`, pod name and uid are baked into the token implicitly